gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+        """
+        sweep_data = self.get_length_sweep(start_length, stop_length, num_pts, group)
+
+        fig = plt.figure()
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s11"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s21"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s12"]) ** 2)
+        plt.plot(sweep_data.loc[:, "length"] / um, abs(sweep_data.loc[:, "s22"]) ** 2)
+        plt.legend(["|S11|^2", "|S21|^2", "|S12|^2", "|S22|^2"])
+        plt.grid("on")
+        plt.xlabel("Length (um)")
+        plt.ylabel("Magnitude")
+        plt.title(f"Length Sweep | Wavelength={self.simulation_settings.wavelength}um")
+        plt.tight_layout()
+        plt.savefig(str(self.dirpath / f"{self.component.name}_length_sweep.png"))
+
+        return fig
+
+    def get_mode_coupling(
+        self, input_mode: int, max_coupled_mode: int = None, group: int = 1
+    ) -> dict[str, pd.DataFrame]:
+        """
+        Get mode coupling coefficients from the input_mode to the max_coupled_mode.
+        Mode numbers start from 1 (fundamental mode).
+
+        Parameters:
+            input_mode: Mode of interest. Energy couples away from this mode to other modes.
+            max_coupled_mode: Maximum mode number to consider; all modes below this mode will be considered as well.
+            group: Group of cells to consider. First group is 0.
+
+        Returns:
+            Dictionary of coupling coefficients vs. position along device:
+            1. forward_sparam - Forward sparam coupling (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
+            2. backward_sparam Backward sparam coupling (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
+            3. overlap - Overlap integrals (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
+            4. position - Position along device (m)
+            | position |
+            | float    |
+        """
+
+        s = self.session
+        ss = self.simulation_settings
+        max_coupled_mode = max_coupled_mode or ss.num_modes
+
+        if s.layoutmode():
+            s.run()
+        s.emepropagate()
+
+        # Get starting point for number of cell and group span
+        cells = [int(val[0]) for val in s.getnamed("EME", "cells")]
+        num_cells = cells[group]
+        group_span = round(s.getnamed("EME", "group spans")[group][0], 15)
+        cell_start = 0
+        for i in range(0, group):
+            cell_start += cells[i]
+
+        ## Extract backward coupling coefficients
+        for i in range(cell_start + 1, num_cells + cell_start + 1):
+            # Get Column 0 Data: S[:,0]
+            # Get Row 0 Data: S[0,:]
+            S = s.getresult(f"EME::Cells::cell_{i}", f"S_{i}_{i+1}")
+
+            # Get coupling based on S parameters and overlap coefficients
+            if i == cell_start + 1:
```

Comment:
you can just initialize `S_bcoupling` to be an empty array, right? since you're concatenating